go-controller/pkg/ovn/service.go

```diff
@@ -57,14 +71,101 @@ func (ovn *Controller) syncServices(services []interface{}) {
+						lb, err := ovn.getGatewayLoadBalancer(gatewayRouter, svcPort.Protocol)
+						if err != nil {
+							klog.Warningf("Service Sync: Gateway router %s does not have load balancer (%v)",
+								gatewayRouter, err)
+							continue
+						}
+						physicalIPs, err := ovn.getGatewayPhysicalIPs(gatewayRouter)
+						for _, physicalIP := range physicalIPs {
+							if err != nil {
+								klog.Warningf("Service Sync: Gateway router %s does not have physical ip (%v)",
+									gatewayRouter, err)
+								continue
+							}
+							name := ovn.generateACLName(lb, physicalIP, svcPort.NodePort)
+							if _, ok := svcRejectACLs[name]; !ok {
+								svcRejectACLs[name] = make(map[string]bool)
+							}
+							svcRejectACLs[name][lb] = hasEndpoints
+						}
+					}
+				}
 			}
 
 			key := util.JoinHostPortInt32(service.Spec.ClusterIP, svcPort.Port)
 			clusterServices[svcPort.Protocol] = append(clusterServices[svcPort.Protocol], key)
-
+			lb, err := ovn.getLoadBalancer(svcPort.Protocol)
+			if err != nil {
+				klog.Warningf("Unable to get existing load balancer from ovn. Reject ACLs may not be synced!")
+			} else {
+				name := ovn.generateACLName(lb, service.Spec.ClusterIP, svcPort.Port)
+				if _, ok := svcRejectACLs[name]; !ok {
+					svcRejectACLs[name] = make(map[string]bool)
+				}
+				svcRejectACLs[name][lb] = hasEndpoints
+			}
 			for _, extIP := range service.Spec.ExternalIPs {
 				key := util.JoinHostPortInt32(extIP, svcPort.Port)
 				lbServices[svcPort.Protocol] = append(lbServices[svcPort.Protocol], key)
+				gateways, _, err := ovn.getOvnGateways()
+				if err != nil {
+					continue
+				}
+				for _, gateway := range gateways {
+					lb, err := ovn.getGatewayLoadBalancer(gateway, svcPort.Protocol)
+					if err != nil {
+						klog.Errorf("Service Sync: Gateway router %s does not have load balancer (%v)",
+							gateway, err)
+						continue
+					}
+					name := ovn.generateACLName(lb, extIP, svcPort.Port)
+					if _, ok := svcRejectACLs[name]; !ok {
+						svcRejectACLs[name] = make(map[string]bool)
+					}
+					svcRejectACLs[name][lb] = hasEndpoints
+				}
+			}
+		}
+	}
+
+	// Get OVN's current reject ACLs. Note, currently only services use reject ACLs.
+	type ovnACLData struct {
+		Data [][]interface{}
+	}
+	data, stderr, err := util.RunOVNNbctl("--columns=name,_uuid", "--format=json", "find", "acl", "action=reject")
+	if err != nil {
+		klog.Errorf("Error while querying ACLs with reject action: %s, %v", stderr, err)
+	} else {
+		x := ovnACLData{}
+		if err := json.Unmarshal([]byte(data), &x); err != nil {
+			klog.Errorf("Unable to get current OVN reject ACLs. Unable to sync reject ACLs!: %v", err)
+		} else if len(x.Data) == 0 {
+			klog.Infof("No reject ACLs currently configured in OVN")
+		} else {
+			for _, entry := range x.Data {
+				// ACL entry format is a slice: [<aclName>, ["_uuid", <uuid>]]
+				if len(entry) != 2 {
+					continue
+				}
+				name := entry[0].(string)
```

Comment:
@trozet I'm a bit worried about cast assertions here if things go wrong with the JSON. Can we have some "ok" stuff here and double-check that our cases are OK so we don't panic?

Reply:
yeah, I didn't even know you can do that in golang. Will add